File: Python/plusOne.py
# Using Linked List
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        n = len(digits)
        abc = ListNode(digits[-1])
        temp = abc

        for i in range(n - 2, -1, -1):
            new_node = ListNode(digits[i])
            temp.next = new_node
            temp = new_node

        temp = abc
        carry = 1
        size = n

        while carry == 1:
            temp.val += carry
            if temp.val >= 10:
                carry = 1
                temp.val %= 10
            else:
                carry = 0
            if temp.next:
                temp = temp.next
            else:
                break

        if carry == 1:
            new_node = ListNode(1)
            temp.next = new_node
            temp = new_node
            size += 1

        ans = [0] * size
        i = size - 1
        while abc:
            ans[i] = abc.val
            abc = abc.next
            i -= 1

        return ans

#Using math
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        n = len(digits)

        for i in range(n - 1, -1, -1):
            if digits[i] + 1 != 10:
                digits[i] += 1
                return digits
            digits[i] = 0

        ans = [1] + [0] * n
        return ans

File: Java/plusone.java
// Using Linkedlist
class ListNode{
    int val;
    ListNode next;
    ListNode(){}
    ListNode(int val){
        this.val= val;
    }
    ListNode(int val, ListNode next){
        this.val= val;
        this.next= next;
    }
}

class Solution {
    public int[] plusOne(int[] digits) {
        int n= digits.length;
        ListNode abc= new ListNode(digits[n-1]);
        ListNode temp= abc;

        for(int i=n-2; i>=0; i--){
            ListNode newNode= new ListNode(digits[i]);
            temp.next= newNode;
            temp= newNode;
        }

        temp= abc;
        int carry= 1, size=n;

        while(carry==1){
            temp.val+=carry;
            if(temp.val>=10){
                carry=1;
                temp.val%=10;
            }else{
                carry=0;
            }
            if(temp.next!=null){
                temp= temp.next;
            }else{
                break;
            }
        }

        if(carry==1){
            ListNode newNode= new ListNode(1);
            temp.next= newNode;
            temp= newNode;
            size+=1;
        } 

        int[] ans= new int[size];
        for(int i=size-1; i>=0 && abc!=null; i--){
            ans[i]= abc.val;
            abc= abc.next;
        }

        return ans;
    }
}

// Using Math
class Solution {
    public int[] plusOne(int[] digits) {

        int n= digits.length;

        for(int i=n-1; i>=0; i--){
            if(digits[i]+1!=10){
                digits[i]+=1;
                return digits;
            }
            digits[i]=0;
        }

        int[] ans= new int[n+1];
        ans[0]=1;

        return ans;
    }
}